lib/vanagon/component.rb

```diff
@@ -350,17 +350,18 @@ def extract_with # rubocop:disable Lint/DuplicateMethods
 
     # Fetches patches if any are provided for the project.
     #
-    # @param workdir [String] working directory to put the patches into
-    def get_patches(workdir)
-      unless @patches.empty?
-        patchdir = File.join(workdir, "patches")
-        FileUtils.mkdir_p(patchdir)
-        @patches.each do |p|
-          target = File.join(patchdir, File.basename(p.path))
-          raise Vanagon::Error, "Duplicate patch files detected, '#{p.path}' would have overwritten '#{target}'! Please ensure patch file names are unique across all components." if File.exist?(target)
-
-          FileUtils.cp(p.path, patchdir)
+    # @param patch_root [String] working directory to put the patches into
+    def get_patches(patch_root)
+      return if @patches.empty?
+      @patches.each do |patch|
+        patch_assembly_path = File.join(patch_root, patch.assembly_path)
+        if File.exist?(patch_assembly_path)
+          raise Vanagon::Error,
+                "Duplicate patch files detected, '#{patch.origin_path}' would have overwritten '#{patch.assembly_path}'"
         end
+        patch_target_directory = File.dirname(patch_assembly_path)
+        FileUtils.mkdir_p(patch_target_directory) unless Dir.exist?(patch_target_directory)
```

Comment:
Don't really need to check if `patch_target_directory` exists, since `mkdir_p` will just no-op if the directory already exists.